src/net_processing.cpp

```diff
@@ -1008,11 +994,11 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans)
         if (nErased > 0) LogPrint(BCLog::MEMPOOL, "Erased %d orphan tx due to expiration\n", nErased);
     }
     FastRandomContext rng;
-    while (mapOrphanTransactions.size() > nMaxOrphans)
-    {
+    while (mapOrphanTransactions.size() > nMaxOrphans) {
         // Evict a random orphan:
-        size_t randompos = rng.randrange(g_orphan_list.size());
-        EraseOrphanTx(g_orphan_list[randompos]->first);
+        auto iter = mapOrphanTransactions.begin();
+        std::advance(iter, rng.randrange(mapOrphanTransactions.size()));
```

Comment:
That's O(n) in the number of orphan transactions. I don't think that's acceptable.

Reply:
Even with given `DEFAULT_MAX_ORPHAN_TRANSACTIONS = 100` ?

Reply:
Benchmarked on the [custom](https://github.com/hebasto/bitcoin/tree/200624-bench) branch:

- master
```
$ ./src/bench/bench_bitcoin -filter="OrphanTxPool"
# Benchmark, evals, iterations, total, min, max, median
OrphanTxPool, 5, 10000, 2.42326, 4.83155e-05, 4.87947e-05, 4.83993e-0
```

- this PR
```
$ ./src/bench/bench_bitcoin -filter="OrphanTxPool"
# Benchmark, evals, iterations, total, min, max, median
OrphanTxPool, 5, 10000, 2.95478, 5.87594e-05, 5.95669e-05, 5.9013e-05
```

Reply:
Thanks for providing actual numbers. I agree it's not so much of a concern in light of those - though I'm also not convinced it's worth the code simplification, especially if we'd ever want/need to increase the maximum number of orphans.

Reply:
[Updated](https://github.com/bitcoin/bitcoin/pull/19374#issuecomment-649536293).